File: flashy/core/flow.py
"""Game flow state machine - defines screen transitions.

This module owns all navigation logic. Screens emit events, and GameFlow
determines what screen to show next. This ensures consistent game flow
across all platforms (TUI, web, iOS).
"""

from dataclasses import dataclass
from enum import Enum, auto
from typing import Any

from flashy.core.levels import get_level
from flashy.core.models import PlayerProgress
from flashy.core.worlds import get_world


class Screen(Enum):
    """All possible screens in the game."""

    PLAYER_SELECT = auto()
    NEW_PLAYER = auto()
    INTRO = auto()
    WORLD_INTRO = auto()
    WORLD_MAP = auto()
    FRIEND_MEET = auto()
    BOSS_INTRO = auto()
    GAMEPLAY = auto()
    RESULT = auto()
    BOSS_VICTORY = auto()
    GAME_COMPLETE = auto()


@dataclass(frozen=True)
class ScreenRequest:
    """Request to show a specific screen with parameters."""

    screen: Screen
    params: dict[str, Any]

    def __repr__(self) -> str:
        return f"ScreenRequest({self.screen.name}, {self.params})"


# --- Events that trigger screen transitions ---


@dataclass(frozen=True)
class GameEvent:
    """Base class for game events."""

    pass


@dataclass(frozen=True)
class AppStarted(GameEvent):
    """App has started."""

    pass


@dataclass(frozen=True)
class PlayerSelected(GameEvent):
    """Player selected from player list."""

    player_name: str
    is_new_player: bool  # True if this is a brand new player


@dataclass(frozen=True)
class NewPlayerCreated(GameEvent):
    """New player profile created."""

    player_name: str


@dataclass(frozen=True)
class IntroDismissed(GameEvent):
    """Intro story screen dismissed."""

    player_name: str


@dataclass(frozen=True)
class WorldIntroDismissed(GameEvent):
    """World intro story screen dismissed."""

    player_name: str
    world_number: int


@dataclass(frozen=True)
class LevelSelected(GameEvent):
    """Level selected from world map."""

    player_name: str
    level_number: int


@dataclass(frozen=True)
class FriendMeetDismissed(GameEvent):
    """Friend meet story screen dismissed."""

    player_name: str
    level_number: int


@dataclass(frozen=True)
class BossIntroDismissed(GameEvent):
    """Boss intro story screen dismissed."""

    player_name: str
    level_number: int


@dataclass(frozen=True)
class LevelCompleted(GameEvent):
    """Level gameplay completed."""

    player_name: str
    level_number: int
    stars: int
    is_new_best: bool
    correct: int
    total: int
    score: int
    best_streak: int


@dataclass(frozen=True)
class ResultContinue(GameEvent):
    """Continue button pressed on result screen."""

    player_name: str
    level_number: int
    stars: int


@dataclass(frozen=True)
class ResultReplay(GameEvent):
    """Replay button pressed on result screen."""

    player_name: str
    level_number: int


@dataclass(frozen=True)
class BossVictoryDismissed(GameEvent):
    """Boss victory screen dismissed."""

    player_name: str
    world_number: int


@dataclass(frozen=True)
class GameCompleteDismissed(GameEvent):
    """Game complete screen dismissed."""

    pass


class GameFlow:
    """Determines screen transitions based on game events.

    This is the single source of truth for game navigation. All platforms
    should use this to ensure consistent behavior.
    """

    def handle(
        self, event: GameEvent, progress: PlayerProgress | None = None
    ) -> ScreenRequest:
        """Handle a game event and return the next screen to show.

        Args:
            event: The game event that occurred.
            progress: Player's current progress (needed for some transitions).

        Returns:
            ScreenRequest indicating which screen to show next.
        """
        if isinstance(event, AppStarted):
            return ScreenRequest(Screen.PLAYER_SELECT, {})

        if isinstance(event, PlayerSelected):
            if event.is_new_player:
                return ScreenRequest(Screen.NEW_PLAYER, {})
            # Existing player - check if they've played before
            if progress and progress.get_highest_unlocked() > 1:
                # Return to world map
                return ScreenRequest(
                    Screen.WORLD_MAP,
                    {"player_name": event.player_name},
                )
            # New-ish player, show intro
            return ScreenRequest(
                Screen.INTRO,
                {"player_name": event.player_name},
            )

        if isinstance(event, NewPlayerCreated):
            return ScreenRequest(
                Screen.INTRO,
                {"player_name": event.player_name},
            )

        if isinstance(event, IntroDismissed):
            return ScreenRequest(
                Screen.WORLD_INTRO,
                {"player_name": event.player_name, "world_number": 1},
            )

        if isinstance(event, WorldIntroDismissed):
            return ScreenRequest(
                Screen.WORLD_MAP,
                {"player_name": event.player_name, "world_number": event.world_number},
            )

        if isinstance(event, LevelSelected):
            level = get_level(event.level_number)
            if level is None:
                return ScreenRequest(
                    Screen.WORLD_MAP,
                    {"player_name": event.player_name},
                )

            # Check for story screens before certain levels
            if level.level_in_world == 6:
                # Friend meet before level 6
                return ScreenRequest(
                    Screen.FRIEND_MEET,
                    {
                        "player_name": event.player_name,
                        "world_number": level.world_number,
                        "level_number": event.level_number,
                    },
                )
            if level.level_in_world == 10:
                # Boss intro before level 10
                return ScreenRequest(
                    Screen.BOSS_INTRO,
                    {
                        "player_name": event.player_name,
                        "world_number": level.world_number,
                        "level_number": event.level_number,
                    },
                )

            # Regular level - go straight to gameplay
            return ScreenRequest(
                Screen.GAMEPLAY,
                {"player_name": event.player_name, "level_number": event.level_number},
            )

        if isinstance(event, FriendMeetDismissed):
            return ScreenRequest(
                Screen.GAMEPLAY,
                {"player_name": event.player_name, "level_number": event.level_number},
            )

        if isinstance(event, BossIntroDismissed):
            return ScreenRequest(
                Screen.GAMEPLAY,
                {"player_name": event.player_name, "level_number": event.level_number},
            )

        if isinstance(event, LevelCompleted):
            return ScreenRequest(
                Screen.RESULT,
                {
                    "player_name": event.player_name,
                    "level_number": event.level_number,
                    "stars": event.stars,
                    "is_new_best": event.is_new_best,
                    "correct": event.correct,
                    "total": event.total,
                    "score": event.score,
                    "best_streak": event.best_streak,
                },
            )

        if isinstance(event, ResultContinue):
            level = get_level(event.level_number)
            if level and level.level_in_world == 10 and event.stars >= 2:
                # Beat the boss! Show victory screen
                return ScreenRequest(
                    Screen.BOSS_VICTORY,
                    {
                        "player_name": event.player_name,
                        "world_number": level.world_number,
                    },
                )

            # Regular continue - back to world map
            if event.stars >= 2:
                next_level = event.level_number + 1
            else:
                next_level = event.level_number
            return ScreenRequest(
                Screen.WORLD_MAP,
                {"player_name": event.player_name, "selected_level": next_level},
            )

        if isinstance(event, ResultReplay):
            return ScreenRequest(
                Screen.GAMEPLAY,
                {"player_name": event.player_name, "level_number": event.level_number},
            )

        if isinstance(event, BossVictoryDismissed):
            next_world = get_world(event.world_number + 1)
            if next_world:
                # More worlds to explore
                return ScreenRequest(
                    Screen.WORLD_INTRO,
                    {
                        "player_name": event.player_name,
                        "world_number": event.world_number + 1,
                    },
                )
            # All worlds complete!
            return ScreenRequest(
                Screen.GAME_COMPLETE,
                {"player_name": event.player_name},
            )

        if isinstance(event, GameCompleteDismissed):
            return ScreenRequest(Screen.PLAYER_SELECT, {})

        # Unknown event - return to player select as fallback
        return ScreenRequest(Screen.PLAYER_SELECT, {})

File: flutter/lib/core/flow.dart
// Game flow state machine - defines screen transitions.
//
// This module owns all navigation logic. Screens emit events, and GameFlow
// determines what screen to show next. This ensures consistent game flow
// across all platforms (web, iOS, Android).
import 'levels.dart';
import 'models.dart';
import 'worlds.dart';

/// All possible screens in the game.
enum Screen {
  playerSelect,
  newPlayer,
  intro,
  worldIntro,
  worldMap,
  friendMeet,
  bossIntro,
  gameplay,
  result,
  bossVictory,
  gameComplete,
}

/// Request to show a specific screen with parameters.
class ScreenRequest {
  const ScreenRequest(this.screen, [this.params = const {}]);

  final Screen screen;
  final Map<String, dynamic> params;

  @override
  String toString() => 'ScreenRequest(${screen.name}, $params)';
}

// --- Events that trigger screen transitions ---

/// Base class for game events.
sealed class GameEvent {
  const GameEvent();
}

/// App has started.
class AppStarted extends GameEvent {
  const AppStarted();
}

/// Player selected from player list.
class PlayerSelected extends GameEvent {
  const PlayerSelected(this.playerName, {required this.isNewPlayer});

  final String playerName;
  final bool isNewPlayer;
}

/// New player profile created.
class NewPlayerCreated extends GameEvent {
  const NewPlayerCreated(this.playerName);

  final String playerName;
}

/// Intro story screen dismissed.
class IntroDismissed extends GameEvent {
  const IntroDismissed(this.playerName);

  final String playerName;
}

/// World intro story screen dismissed.
class WorldIntroDismissed extends GameEvent {
  const WorldIntroDismissed(this.playerName, {required this.worldNumber});

  final String playerName;
  final int worldNumber;
}

/// Level selected from world map.
class LevelSelected extends GameEvent {
  const LevelSelected(this.playerName, {required this.levelNumber});

  final String playerName;
  final int levelNumber;
}

/// Friend meet story screen dismissed.
class FriendMeetDismissed extends GameEvent {
  const FriendMeetDismissed(this.playerName, {required this.levelNumber});

  final String playerName;
  final int levelNumber;
}

/// Boss intro story screen dismissed.
class BossIntroDismissed extends GameEvent {
  const BossIntroDismissed(this.playerName, {required this.levelNumber});

  final String playerName;
  final int levelNumber;
}

/// Level gameplay completed.
class LevelCompleted extends GameEvent {
  const LevelCompleted({
    required this.playerName,
    required this.levelNumber,
    required this.stars,
    required this.isNewBest,
    required this.correct,
    required this.total,
    required this.score,
    required this.bestStreak,
  });

  final String playerName;
  final int levelNumber;
  final int stars;
  final bool isNewBest;
  final int correct;
  final int total;
  final int score;
  final int bestStreak;
}

/// Continue button pressed on result screen.
class ResultContinue extends GameEvent {
  const ResultContinue({
    required this.playerName,
    required this.levelNumber,
    required this.stars,
  });

  final String playerName;
  final int levelNumber;
  final int stars;
}

/// Replay button pressed on result screen.
class ResultReplay extends GameEvent {
  const ResultReplay({required this.playerName, required this.levelNumber});

  final String playerName;
  final int levelNumber;
}

/// Boss victory screen dismissed.
class BossVictoryDismissed extends GameEvent {
  const BossVictoryDismissed({
    required this.playerName,
    required this.worldNumber,
  });

  final String playerName;
  final int worldNumber;
}

/// Game complete screen dismissed.
class GameCompleteDismissed extends GameEvent {
  const GameCompleteDismissed();
}

/// Determines screen transitions based on game events.
///
/// This is the single source of truth for game navigation. All platforms
/// should use this to ensure consistent behavior.
class GameFlow {
  /// Handle a game event and return the next screen to show.
  ///
  /// [event] The game event that occurred.
  /// [progress] Player's current progress (needed for some transitions).
  ///
  /// Returns ScreenRequest indicating which screen to show next.
  ScreenRequest handle(GameEvent event, [PlayerProgress? progress]) {
    switch (event) {
      case AppStarted():
        return const ScreenRequest(Screen.playerSelect);

      case PlayerSelected(:final playerName, :final isNewPlayer):
        if (isNewPlayer) {
          return const ScreenRequest(Screen.newPlayer);
        }
        // Existing player - check if they've played before
        if (progress != null && progress.getHighestUnlocked() > 1) {
          return ScreenRequest(Screen.worldMap, {'player_name': playerName});
        }
        // New-ish player, show intro
        return ScreenRequest(Screen.intro, {'player_name': playerName});

      case NewPlayerCreated(:final playerName):
        return ScreenRequest(Screen.intro, {'player_name': playerName});

      case IntroDismissed(:final playerName):
        return ScreenRequest(Screen.worldIntro, {
          'player_name': playerName,
          'world_number': 1,
        });

      case WorldIntroDismissed(:final playerName, :final worldNumber):
        return ScreenRequest(Screen.worldMap, {
          'player_name': playerName,
          'world_number': worldNumber,
        });

      case LevelSelected(:final playerName, :final levelNumber):
        final level = getLevel(levelNumber);
        if (level == null) {
          return ScreenRequest(Screen.worldMap, {'player_name': playerName});
        }

        // Check for story screens before certain levels
        if (level.levelInWorld == 6) {
          // Friend meet before level 6
          return ScreenRequest(Screen.friendMeet, {
            'player_name': playerName,
            'world_number': level.worldNumber,
            'level_number': levelNumber,
          });
        }
        if (level.levelInWorld == 10) {
          // Boss intro before level 10
          return ScreenRequest(Screen.bossIntro, {
            'player_name': playerName,
            'world_number': level.worldNumber,
            'level_number': levelNumber,
          });
        }

        // Regular level - go straight to gameplay
        return ScreenRequest(Screen.gameplay, {
          'player_name': playerName,
          'level_number': levelNumber,
        });

      case FriendMeetDismissed(:final playerName, :final levelNumber):
        return ScreenRequest(Screen.gameplay, {
          'player_name': playerName,
          'level_number': levelNumber,
        });

      case BossIntroDismissed(:final playerName, :final levelNumber):
        return ScreenRequest(Screen.gameplay, {
          'player_name': playerName,
          'level_number': levelNumber,
        });

      case LevelCompleted(
          :final playerName,
          :final levelNumber,
          :final stars,
          :final isNewBest,
          :final correct,
          :final total,
          :final score,
          :final bestStreak
        ):
        return ScreenRequest(Screen.result, {
          'player_name': playerName,
          'level_number': levelNumber,
          'stars': stars,
          'is_new_best': isNewBest,
          'correct': correct,
          'total': total,
          'score': score,
          'best_streak': bestStreak,
        });

      case ResultContinue(:final playerName, :final levelNumber, :final stars):
        final level = getLevel(levelNumber);
        if (level != null && level.levelInWorld == 10 && stars >= 2) {
          // Beat the boss! Show victory screen
          return ScreenRequest(Screen.bossVictory, {
            'player_name': playerName,
            'world_number': level.worldNumber,
          });
        }

        // Regular continue - back to world map
        final nextLevel = stars >= 2 ? levelNumber + 1 : levelNumber;
        return ScreenRequest(Screen.worldMap, {
          'player_name': playerName,
          'selected_level': nextLevel,
        });

      case ResultReplay(:final playerName, :final levelNumber):
        return ScreenRequest(Screen.gameplay, {
          'player_name': playerName,
          'level_number': levelNumber,
        });

      case BossVictoryDismissed(:final playerName, :final worldNumber):
        final nextWorld = getWorld(worldNumber + 1);
        if (nextWorld != null) {
          // More worlds to explore
          return ScreenRequest(Screen.worldIntro, {
            'player_name': playerName,
            'world_number': worldNumber + 1,
          });
        }
        // All worlds complete!
        return ScreenRequest(Screen.gameComplete, {'player_name': playerName});

      case GameCompleteDismissed():
        return const ScreenRequest(Screen.playerSelect);
    }
  }
}
